### PR TITLE
Add Liveness and Readiness probes to Mirror Maker

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

This Pr adds liveness and readiness probes for Mirror Maker ... which we surprisingly didn't had yet :-/.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java [hunk 1]
```diff
@@ -369,6 +381,15 @@ public Deployment generateDeployment(Map<String, String> annotations, boolean is
                 .withCommand("/opt/kafka/kafka_mirror_maker_run.sh")
                 .withEnv(getEnvVars())
                 .withPorts(getContainerPortList())
+                .withLivenessProbe(ModelUtils.newProbeBuilder(livenessProbeOptions)
+                        .withNewExec()
+                        .withCommand("/opt/kafka/kafka_mirror_maker_liveness.sh")
+                        .endExec().build())
+                .withReadinessProbe(ModelUtils.newProbeBuilder(readinessProbeOptions)
+                        .withNewExec()
+                        // The kafka-agent will create /var/opt/kafka/kafka-ready in the container
```
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java]:
```suggestion
                        // The kafka-mirror-maker agent will create /tmp/mirror-maker-ready in the container
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java [hunk 2]
```diff
@@ -434,8 +474,14 @@ public Deployment generateDeployment(Map<String, String> annotations, boolean is
...
         }
+    }
 
-        /** producer */
+    /**
+     * Sets the producer related environment variables in the provided List.
+     *
+     * @param varList   List with environment variables
+     */
+    private void setProducerEnvVars(List<EnvVar> varList) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java]: Can we rename this `addProducerEnvVars` ?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java [hunk 3]
```diff
@@ -406,6 +427,25 @@ public Deployment generateDeployment(Map<String, String> annotations, boolean is
...
+
+        return varList;
+    }
+
+    /**
+     * Sets the consumer related environment variables in the provided List.
+     *
+     * @param varList   List with environment variables
+     */
+    private void setConsumerEnvVars(List<EnvVar> varList) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java]: Can we rename this `addConsumerEnvVars` ?

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests testcase=UserST#testUserWithNameMoreThan64Chars,UserST#testBigAmountOfTlsUsers

File: documentation/book/appendix_crds.adoc [hunk 4]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
-|Property            |Description
-|replicas     1.2+<.<|The number of pods in the `Deployment`.
+|Property               |Description
+|replicas        1.2+<.<|The number of pods in the `Deployment`.
 |integer
-|image        1.2+<.<|The docker image for the pods.
+|image           1.2+<.<|The docker image for the pods.
 |string
-|whitelist    1.2+<.<|List of topics which are included for mirroring. This option allows any regular expression using Java-style regular expressions. Mirroring two topics named A and B can be achieved by using the whitelist `'A\|B'`. Or, as a special case, you can mirror all topics using the whitelist '*'. Multiple regular expressions separated by commas can be specified as well.
+|whitelist       1.2+<.<|List of topics which are included for mirroring. This option allows any regular expression using Java-style regular expressions. Mirroring two topics named A and B can be achieved by using the whitelist `'A\|B'`. Or, as a special case, you can mirror all topics using the whitelist '*'. Multiple regular expressions separated by commas can be specified as well.
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|whitelist       1.2+<.<|List of topics which are included for mirroring. This option allows any regular expression using Java-style regular expressions. Mirroring two topics named A and B can be achieved by using the whitelist `'A\|B'`. Or, as a special case, you can mirror all topics using the whitelist `*`. Multiple regular expressions separated by commas can be specified as well.
```

File: documentation/book/appendix_crds.adoc [hunk 5]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
+|consumer        1.2+<.<|Configuration of source cluster.
 |xref:type-KafkaMirrorMakerConsumerSpec-{context}[`KafkaMirrorMakerConsumerSpec`]
-|producer     1.2+<.<|Configuration of target cluster.
+|producer        1.2+<.<|Configuration of target cluster.
 |xref:type-KafkaMirrorMakerProducerSpec-{context}[`KafkaMirrorMakerProducerSpec`]
-|resources    1.2+<.<|Resource constraints (limits and requests).
+|resources       1.2+<.<|Resource constraints (limits and requests).
 |xref:type-ResourceRequirements-{context}[`ResourceRequirements`]
-|affinity     1.2+<.<|*The property `affinity` has been deprecated. This feature should now be configured at path `spec.template.pod.affinity`.* The pod's affinity rules.See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#affinity-v1-core[core/v1 affinity].
+|affinity        1.2+<.<|*The property `affinity` has been deprecated. This feature should now be configured at path `spec.template.pod.affinity`.* The pod's affinity rules.See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#affinity-v1-core[core/v1 affinity].
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|affinity        1.2+<.<|*The property `affinity` has been deprecated. This feature should now be configured at path `spec.template.pod.affinity`.* The pod's affinity rules. See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#affinity-v1-core[core/v1 affinity].
```

File: documentation/book/appendix_crds.adoc [hunk 6]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
 
 
 |https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#affinity-v1-core[Affinity]
-|tolerations  1.2+<.<|*The property `tolerations` has been deprecated. This feature should now be configured at path `spec.template.pod.tolerations`.* The pod's tolerations.See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#toleration-v1-core[core/v1 toleration].
+|tolerations     1.2+<.<|*The property `tolerations` has been deprecated. This feature should now be configured at path `spec.template.pod.tolerations`.* The pod's tolerations.See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#toleration-v1-core[core/v1 toleration].
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|tolerations     1.2+<.<|*The property `tolerations` has been deprecated. This feature should now be configured at path `spec.template.pod.tolerations`.* The pod's tolerations. See external documentation of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#toleration-v1-core[core/v1 toleration].
```

File: documentation/book/appendix_crds.adoc [hunk 7]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
 
 
 |https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#toleration-v1-core[Toleration] array
-|jvmOptions   1.2+<.<|JVM Options for pods.
+|jvmOptions      1.2+<.<|JVM Options for pods.
 |xref:type-JvmOptions-{context}[`JvmOptions`]
-|logging      1.2+<.<|Logging configuration for Mirror Maker. The type depends on the value of the `logging.type` property within the given object, which must be one of [inline, external].
+|logging         1.2+<.<|Logging configuration for Mirror Maker. The type depends on the value of the `logging.type` property within the given object, which must be one of [inline, external].
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|logging         1.2+<.<|Logging configuration for Mirror Maker. The type depends on the value of the `logging.type` property within the given object, which must be `inline` or `external`.
```

File: documentation/book/appendix_crds.adoc [hunk 8]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
 |xref:type-InlineLogging-{context}[`InlineLogging`], xref:type-ExternalLogging-{context}[`ExternalLogging`]
-|metrics      1.2+<.<|The Prometheus JMX Exporter configuration. See {JMXExporter} for details of the structure of this configuration.
+|metrics         1.2+<.<|The Prometheus JMX Exporter configuration. See {JMXExporter} for details of the structure of this configuration.
 |map
-|template     1.2+<.<|Template for Kafka Mirror Maker resources. The template allows users to specify how is the `Deployment` and `Pods` generated.
+|template        1.2+<.<|Template for Kafka Mirror Maker resources. The template allows users to specify how is the `Deployment` and `Pods` generated.
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|template        1.2+<.<|Template to specify how Kafka Mirror Maker resources, `Deployment` and `Pods`, are generated.
```

File: documentation/book/appendix_crds.adoc [hunk 9]
```diff
@@ -1736,36 +1736,40 @@ Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
...
 |xref:type-KafkaMirrorMakerTemplate-{context}[`KafkaMirrorMakerTemplate`]
-|version      1.2+<.<|The Kafka Mirror Maker version. Defaults to {DefaultKafkaVersion}. Consult the user documentation to understand the process required to upgrade or downgrade the version.
+|livenessProbe   1.2+<.<|Pod liveness checking.
+|xref:type-Probe-{context}[`Probe`]
+|readinessProbe  1.2+<.<|Pod readiness checking.
+|xref:type-Probe-{context}[`Probe`]
+|version         1.2+<.<|The Kafka Mirror Maker version. Defaults to {DefaultKafkaVersion}. Consult the user documentation to understand the process required to upgrade or downgrade the version.
```
Comment (PaulRMellor) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|version         1.2+<.<|The Kafka Mirror Maker version. Defaults to {DefaultKafkaVersion}. Consult the Mirror Maker documentation to understand the process required to upgrade or downgrade the version.
```

**Comment by scholzj** [comment]
@PaulRMellor Thanks for the comments. The points you raised are not related to this PR really. The only reason why GitHub marked them as changed is that the API docs generator changed the whole table. Is it ok for you to merge this PR and open a separate on for these changes? It will be much cleaner approach.

**Comment by PaulRMellor** [comment]
> @PaulRMellor Thanks for the comments. The points you raised are not related to this PR really. The only reason why GitHub marked them as changed is that the API docs generator changed the whole table. Is it ok for you to merge this PR and open a separate on for these changes? It will be much cleaner approach.

Sure Jakub.

**Pull request merged**